dlc/src/main/scala/org/bitcoins/dlc/DLCMessage.scala

```diff
@@ -2,37 +2,502 @@ package org.bitcoins.dlc
-import org.bitcoins.core.protocol.Bech32Address
+import org.bitcoins.core.number.UInt32
+import org.bitcoins.core.protocol.BlockStamp.BlockTime
 import org.bitcoins.core.protocol.transaction.{
+  OutputReference,
   TransactionOutPoint,
   TransactionOutput
 }
+import org.bitcoins.core.protocol.{
+  Bech32Address,
+  BitcoinAddress,
+  NetworkElement
+}
+import org.bitcoins.core.psbt.InputPSBTRecord.PartialSignature
+import org.bitcoins.core.util.{CryptoUtil, Factory}
 import org.bitcoins.core.wallet.fee.SatoshisPerVirtualByte
+import scodec.bits.ByteVector
+import ujson._
+
+import scala.annotation.tailrec
+import scala.collection.{mutable, GenTraversableOnce}
 
-sealed trait DLCMessage
+sealed trait DLCMessage {
+  def toJson: Value
+  def toJsonStr: String = toJson.toString()
+}
 
 object DLCMessage {
+
+  // TODO: will need to be changed when this is standardized
+  def calcEventId(
+      oracleInfo: OracleInfo,
+      contractInfo: ContractInfo,
+      timeouts: DLCTimeouts): Sha256DigestBE = {
+    CryptoUtil
+      .sha256(oracleInfo.bytes ++ contractInfo.bytes ++ timeouts.bytes)
+      .flip
+  }
+
   case class OracleInfo(pubKey: ECPublicKey, rValue: ECPublicKey)
+      extends NetworkElement {
+
+    override def bytes: ByteVector = pubKey.bytes ++ rValue.bytes
+  }
+
+  object OracleInfo extends Factory[OracleInfo] {
+
+    val empty: OracleInfo = OracleInfo(ByteVector.low(66))
+
+    override def fromBytes(bytes: ByteVector): OracleInfo = {
+      require(bytes.size == 66, s"OracleInfo is only 66 bytes, got $bytes")
+
+      val pubkey = ECPublicKey(bytes.take(33))
+      val rValue = ECPublicKey(bytes.drop(33))
+
+      // TODO: validate rValue once the libsecp functionality is added
+
+      OracleInfo(pubkey, rValue)
+    }
+  }
+
+  case class ContractInfo(outcomeValueMap: Map[Sha256DigestBE, Satoshis])
+      extends NetworkElement {
+    override def bytes: ByteVector = {
+      outcomeValueMap.foldLeft(ByteVector.empty) {
+        case (vec, (digest, sats)) => vec ++ digest.bytes ++ sats.bytes
+      }
+    }
+    // Give common Map methods so that users don't have to reach in every time
+
+    def head: (Sha256DigestBE, Satoshis) = outcomeValueMap.head
+    def last: (Sha256DigestBE, Satoshis) = outcomeValueMap.last
+
+    def tail: Map[Sha256DigestBE, Satoshis] = outcomeValueMap.tail
+    def init: Map[Sha256DigestBE, Satoshis] = outcomeValueMap.init
+
+    def keys: Iterable[Sha256DigestBE] = outcomeValueMap.keys
+    def values: Iterable[Satoshis] = outcomeValueMap.values
 
+    def toVector: Vector[(Sha256DigestBE, Satoshis)] = outcomeValueMap.toVector
+
+    def foldLeft[B](z: B)(op: (B, (Sha256DigestBE, Satoshis)) => B): B =
+      outcomeValueMap.foldLeft(z)(op)
+
+    def map[B](f: ((Sha256DigestBE, Satoshis)) => B): Iterable[B] =
+      outcomeValueMap.map(f)
+
+    def flatMap[B](
+        f: ((Sha256DigestBE, Satoshis)) => GenTraversableOnce[B]): Iterable[B] =
+      outcomeValueMap.flatMap(f)
+  }
```

Comment:
I think in the future (doesn't have to be this PR), it would be nice to have a trait for `case class`es that wrap `Vector` and `Map` to make it so that this boiler-plate doesn't have to be pasted in all the wrappers. Essentially something like
```
trait VectorWrapper[+T] {
  def vec: Vector[T]
  def head: T = vec.head
  def tail: Vector[T] = vec.tail
  def map[U](func: T => U): Vector[U] = vec.map(func)
  ...
}

trait MapWrapper[T, +U] extends VectorWrapper[(T, U)] {
  def mapVal: Map[T, U]
  override def vec: Vector[(T, U)] = mapVal.toVector
  def keys: Vector[T] = mapVal.keys.toVector
  ...
}
```
so that we could then have:
```
case class ContractInfo(override val mapVal: Map[Sha256DigestBE, Satoshis]) extends MapWrapper[Sha256DigestBE, Satoshis]
```

Reply:
#1131